enhancements/local-storage/subprovisioner-integration-into-lvms.md

```diff
@@ -0,0 +1,255 @@
+    // Name of the class, the VG and possibly the storageclass.
+    // Validations to confirm that this field can be used as metadata.name field in storageclass
+    // ref: https://github.com/kubernetes/apimachinery/blob/de7147/pkg/util/validation/validation.go#L209
+    // +kubebuilder:validation:MaxLength=245
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:Pattern="^[a-z0-9]([-a-z0-9]*[a-z0-9])?$"
+    Name string `json:"name,omitempty"`
+    
+    // DeviceSelector is a set of rules that should match for a device to be included in the LVMCluster
+    // +optional
+    DeviceSelector *DeviceSelector `json:"deviceSelector,omitempty"`
+    
+    // NodeSelector chooses nodes on which to create the deviceclass
+    // +optional
+    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+    
+    // ThinPoolConfig contains configurations for the thin-pool
+    // +optional
+    ThinPoolConfig *ThinPoolConfig `json:"thinPoolConfig,omitempty"`
+    
+    // Default is a flag to indicate whether the device-class is the default.
+    // This will mark the storageClass as default.
+    // +optional
+    Default bool `json:"default,omitempty"`
+    
+    // FilesystemType sets the filesystem the device should use.
+    // For shared deviceClasses, this field must be set to "" or none.
+    // +kubebuilder:validation:Enum=xfs;ext4;none;""
+    // +kubebuilder:default=xfs
+    // +optional
+    FilesystemType DeviceFilesystemType `json:"fstype,omitempty"`
+    
+    // Policy defines the policy for the deviceClass.
+    // TECH PREVIEW: shared will allow accessing the deviceClass from multiple nodes.
+    // The deviceClass will then be configured via shared volume group.
+    // +optional	  
+    // +kubebuilder:validation:Enum=shared;local
++   DeviceAccessPolicy DeviceAccessPolicy `json:"deviceAccessPolicy,omitempty"`
+  }
+
+  type ThinPoolConfig struct {
+    // Name of the thin pool to be created. Will only be used for node-local storage, 
+    // since shared volume groups will create a thin pool with the same name as the volume group.
+    // +kubebuilder:validation:Required
+    // +required
+    Name string `json:"name"`
+    
+    // SizePercent represents percentage of space in the volume group that should be used
+    // for creating the thin pool.
+    // +kubebuilder:default=90
+    // +kubebuilder:validation:Minimum=10
+    // +kubebuilder:validation:Maximum=90
+    SizePercent int `json:"sizePercent,omitempty"`
+    
+    // OverProvisionRatio is the factor by which additional storage can be provisioned compared to
+    // the available storage in the thin pool. Only applicable for node-local storage.
+    // +kubebuilder:validation:Minimum=1
+    // +kubebuilder:validation:Required
+    // +required
+    OverprovisionRatio int `json:"overprovisionRatio"`
+  }
+
+  // DeviceSelector specifies the list of criteria that have to match before a device is assigned
+  type DeviceSelector struct {
```

Comment:
The deviceSelector is very limited in nature and paths / optionalPaths are expected to remain constant mostly (with some exceptions). For this reason one is guided to use immutable disk identifiers. Not too familiar with multipath devices here, so maybe give me a sample configuration and I can check if it would work.

Reply:
Okay, thanks for clarifying. On the Subprovisioner side we'll need to investigate the consequences of using multiple physical disks since we haven't considered it before.

Reply:
isnt that the default for a multi path LUN?

Reply:
I don't think so. Physical LUNs are combined into a single mpath device with dm-multipath. LVM uses just the single mpath device and doesn't see multiple devices.

Anyway, we have successfully tested both multiple devices and multipath now! :+1:

Reply:
> the single mpath device

I think this is a bit misleading because a multipath devices can have 1 active and multiple passive paths which can fail over. I guess Im wondering how this device discovery and check would work?

Reply:
Right now the node admin needs to configure dm-multipath and created an LVM Volume Group that uses the mpath device. The admin creates a StorageClass with the Volume Group name so that Subprovisioner will use that Volume Group.

In a scenario where the operator is responsible for creating the Volume Group, the node admin could configure  dm-multipath manually and then specify a DeviceSelector to look for the mpath device. The operator would then create the Volume Group that uses the mpath device.

So in the end, the node admin still needs to set up dm-multipath manually. Do you think that approach is reasonable, or do you want the operator to configure dm-multipath automatically?

Reply:
In my view, There is no such thing as "configuring dm-multipath manually" with an operator. If that is configured, then by CoreOS layering, or MachineConfig which would have to be done from the operator. That being said, I believe that we can outsource the MachineConfig creation in a Prerequisite steps into the Docs, WDYT?

Reply:
Yes, you are right. "Manually" is the wrong word. Cluster admins should enable dm-multipath at OpenShift installation time or using MachineConfig:
- https://docs.openshift.com/container-platform/4.15/installing/installing_bare_metal/installing-bare-metal.html#rhcos-enabling-multipath_installing-bare-metal
- https://docs.openshift.com/container-platform/4.15/post_installation_configuration/machine-configuration-tasks.html#rhcos-enabling-multipath-day-2_post-install-machine-configuration-tasks